Lazy Profiling + Debug "lazy-trace" (wip)

SkySkimmer opened the pull request:
also Ccnv Profiling

SkySkimmer committed:
Lazy Profiling + Debug "lazy-trace"

SkySkimmer commented:
@coqbot bench

SkySkimmer commented:
@coqbot bench

Janno commented:
This is cool! I've been trying to see if I can learn anything from the output. The output of `lazy-trace` seems to contain a lot of useful information but it is very very hard to parse without building a proper tool for it. The automatically inserted line breaks really work against my usual approach of throwing a bunch of `sed` and `rg` at it. It would be very helpful to have a json based output (or anything else that offers tools comparable to `jq`). 

The way I envision using this is to pose queries such as "how match steps happen directly inside this function", "how often do we reduce function A in function B", "is there a transitive call chain from function C to function D" etc. 

I have managed to at least get the entire stack on one line. I guess I could now use a bunch more grepping to answer these kinds of questions.
```bash
cat /tmp/trace-qed.out | rg -U --multiline-dotall '<infomsg>.*?Debug:.*?\[lazy-trace\].*?(.*?)</infomsg>' -r '$1' -o --json | sed -re 's,\\n,,g' | sed -re 's,  +, ,g' | rg '<infomsg>.*?Debug:.*?\[lazy-trace\].*?(.*?)</infomsg>' -r '$1' -o | sed -re 's,^ ,,'
```

It might also be nice to expose control over the counts and traces to clients. I have calls to kernel reduction that really should be grouped when it comes to statistics and traces. It would also be very easy to do some kind of processing on the traces directly in OCaml.

I've also run performance tests on BlueRock code. Somewhat expectedly, it impacts our generated files (one huge `Definition`) more than our proofs (lots of time in tactic code).

| Relative | Master   | MR       | Change   | Filename
|---------:|---------:|---------:|---------:|----------
|   +0.79% |  35975.3 |  36260.5 |   +285.1 | cpp2v-generated
|   +0.16% |  87992.2 |  88129.6 |   +137.5 | other
|   +0.34% | 123967.5 | 124390.1 |   +422.6 | total

Some more thoughts:

Mtac2 had tracing that looked quite similar to this approach. (@ppedrot helped me figure out how to time things correctly in the crazy call-by-name semantics of Mtac2.) The tracing was limited to just delta steps and Mtac2 primitives because almost everything else is handled directly by the kernel. Most importantly, it also recorded time stamps which I used it to generate flame graphs (see very old example below). Those were very helpful to figure out where Mtac2 spends its time. I wonder if it would be too much overhead to record timestamps in Coq. Unfortunately it is possible that getting the current time at a useful resolution adds noticeable overhead that could skew the results.

<details><summary>Example SVG flame graph of an Mtac2 program</summary>

![heap_lang](https://github.com/user-attachments/assets/9928071d-f476-47a5-b441-e85f055d9ab2)

</details>

On a wilder note, and maybe as an alternative to recording timestamps that won't skew the results: `gdb` has support for stopping programs regularly and poking around in their state. It can be used to take stack samples and it could also be used to sample the kernel trace added by this MR *if* the trace was exposed at a fixed or somehow discoverable address. Finding the trace is probably the hardest part with this idea. Writing a bit of `gdb` script code to traverse an OCaml list should be doable. Translating the names to strings might be more difficult but I am positive that some solution could be figured out.

The reason I am suggesting this is that if we can get the timestamps to agree with `perf`'s timestamps (using `CLOCK_MONOTONIC`, probably) we could finally get performance results for `Qed` that align the OCaml code with the Coq functions being reduced.

SkySkimmer commented:
>I've also run performance tests on BlueRock code. 

I guess this is with the debug flags off ie only measures the cost when not using the feature?

Janno commented:
> I guess this is with the debug flags off ie only measures the cost when not using the feature?

Yes, this is just using the current code of the MR. If you want I can tweak the code to make it perform the work of recording without actually printing anything (because that would probably wreck our CI logs) and rerun the measurements.

SkySkimmer commented:
@coqbot bench

@Janno I pushed

Janno commented:
| Relative | Master   | MR       | Change   | Filename
|---------:|---------:|---------:|---------:|----------
|   +5.71% |  35975.3 |  38027.8 |  +2052.4 | cpp2v-generated
|   +0.89% |  87991.9 |  88774.7 |   +782.8 | other
|   +2.29% | 123967.2 | 126802.4 |  +2835.2 | total

SkySkimmer commented:
seems like that would do a lot of extra allocation

Janno commented:
I tried collecting events in a list and it seems slower than the hashtable. Despite the fact that I fixed the problem with work happening in `print_recorded_steps` when nothing is supposed to be printed. 

Here's my change: https://github.com/coq/coq/commit/8447d30f5983612aa11aff2ab21194914caa4cfc. 

Results on our code base (relative to master without this MR, not relative to the previous version of the MR):

| Relative | Master   | MR       | Change   | Filename
|---------:|---------:|---------:|---------:|----------
|   +6.73% |  35975.4 |  38397.4 |  +2422.0 | cpp2v-generated
|   +1.70% |  87991.9 |  89488.8 |  +1496.9 | other
|   +3.16% | 123967.2 | 127886.2 |  +3919.0 | total

SkySkimmer commented:
>`  let create : t = ref []`

that seems wrong, you're accumulating to a global ref instead of creating one for the given `clos_tab`

SkySkimmer commented:
>I think the fact that this call is unconditional really messes with the latest commit. 

fixed

@coqbot bench

SkySkimmer commented:
>It's still pretty bad, though, absolutely speaking.

The standard way to go faster when profiling would be to record less than 100% of the steps

SkySkimmer commented:
@coqbot bench

SkySkimmer committed:
Also apply sample rate to lazy timings

Unlike the step counts, the timings are measured such that they don't
get scaled by the sample rate (ie with rate 1/10, 100 steps becomes
10ish steps but 1s is still 1s).

There may be some distortion of the results, but other that dropping
very short operations from the table (eg the delta from unfolding a
`Definition foo := bar`) it experimentally seems to produce reasonable
results.

SkySkimmer commented:
@coqbot bench
